man/man3/MobileOperator.3
.TH "MobileOperator" 3 "Thu May 20 2021" "Simulator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MobileOperator
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <MobileOperator\&.h>\fP
.PP
Inherits \fBAgent\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBMobileOperator\fP (const \fBMap\fP *m, const unsigned long id, const \fBClock\fP *clock, const char *name, const double probMobilePhone, const string &outputDir)"
.br
.ti -1c
.RI "virtual \fB~MobileOperator\fP ()"
.br
.ti -1c
.RI "const string \fBtoString\fP (bool detailed=false) const override"
.br
.ti -1c
.RI "const string \fBgetMNOName\fP () const"
.br
.ti -1c
.RI "const double \fBgetProbMobilePhone\fP () const"
.br
.ti -1c
.RI "ofstream & \fBgetAntennaCellsFile\fP ()"
.br
.ti -1c
.RI "ofstream & \fBgetSignalFile\fP ()"
.br
.ti -1c
.RI "void \fBwriteSignalFileHeader\fP ()"
.br
.ti -1c
.RI "const string & \fBgetOutputDir\fP () const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static const string \fBgetHeader\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "const string \fBm_name\fP"
.br
.ti -1c
.RI "const double \fBm_probMobilePhone\fP"
.br
.ti -1c
.RI "ofstream \fBm_antennaCellsFileName\fP"
.br
.ti -1c
.RI "ofstream \fBm_signalMeasureFileName\fP"
.br
.ti -1c
.RI "const string & \fBm_outputDir\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
This class represents a Mobile Operator company\&. Currently a simulation can be run with 1 or 2 mobile operators\&. A mobile operator owns a set of antennas and has a set of mobile phone subscribed\&. It has a name read from the simulation configuration file and two ofstream members representing the files where the cell boundaries and signal strength/dominance are saved for all antennas belonging to this mobile operator\&. It also has a member representing the probability of a person to have a mobile phone subscribed to this mobile network operator\&. The value of this probability is read from the simulation configuration file\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "MobileOperator::MobileOperator (const \fBMap\fP * m, const unsigned long id, const \fBClock\fP * clock, const char * name, const double probMobilePhone, const string & outputDir)"
Constructor of this class\&. It builds a \fBMobileOperator\fP object with the characteristics provided by user through the parameters of the constructor\&. In the process of creating the object, the constructor method builds a string representing the name of the file where the coverage area of all the antennas (cell boundary) that belong to this \fBMobileOperator\fP are saved in \&.csv format, then open this file and writes the header of the file\&. The name of this file is built by concatenating 'AntennaCells_' with the name of the Mobile Operator\&. A line of this file contains the antenna id followed by a \fCwkt\fP text representing the coverage area of antenna\&. It also builds another string representing the name of a file where the signal strength/dominance values for all antennas that belong to this \fBMobileOperator\fP are saved\&. The signal strength/dominance is computed in the center of each tile of the \fBGrid\fP object set for a simulation\&. The file is opened and then it writes the header of the file\&. The name of this file is built by concatenating 'SignalMeasure_' with the name of the Mobile Operator\&. A line of this file contains the antenna id followed by a set of values for the signal strength/dominance computed in the center of each tile of the grid\&. 
.PP
\fBParameters\fP
.RS 4
\fIm\fP a pointer to a \fBMap\fP object of the simulation\&. 
.br
\fIid\fP the id of the \fBMobileOperator\fP object\&. 
.br
\fIclock\fP a pointer to a \fBClock\fP object used for a simulation\&. 
.br
\fIname\fP the name of the Mobile Operator\&. 
.br
\fIprobMobilePhone\fP represents the probability for a person to have a mobile phone subscribed to this company\&. 
.br
\fIoutputDir\fP a string representing the name of the directory where the output files are written\&. 
.RE
.PP

.SS "virtual MobileOperator::~MobileOperator ()\fC [virtual]\fP"
Default destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "ofstream& MobileOperator::getAntennaCellsFile ()"
Returns the file where the coverage area of all the antennas that belong to this \fBMobileOperator\fP are saved in \&.csv format\&. 
.PP
\fBReturns\fP
.RS 4
an ofstream object representing the file where the coverage area of all the antennas that belong to this \fBMobileOperator\fP are saved in \&.csv format\&. 
.RE
.PP

.SS "static const string MobileOperator::getHeader ()\fC [static]\fP"
Returns the header of the string representation of \fBMobileOperator\fP objects\&. 
.PP
\fBReturns\fP
.RS 4
the header of the string representation of \fBMobileOperator\fP objects\&. 
.RE
.PP

.SS "const string MobileOperator::getMNOName () const"
The name of the Mobile Operator\&. This name is read from the simulation configuration file and it is provided as a parameter to the constructor of the class\&. 
.PP
\fBReturns\fP
.RS 4
the name of the Mobile Operator 
.RE
.PP

.SS "const string& MobileOperator::getOutputDir () const"
Returns the name of the folder where the output files are saved\&. 
.PP
\fBReturns\fP
.RS 4
the name of the folder where the output files are saved\&. 
.RE
.PP

.SS "const double MobileOperator::getProbMobilePhone () const"
The probability for a person to have a mobile phone subscribed to this company\&. It is provided as a parameter to the constructor of the class\&. 
.PP
\fBReturns\fP
.RS 4
the probability for a person to have a mobile phone subscribed to this company\&. 
.RE
.PP

.SS "ofstream& MobileOperator::getSignalFile ()"
Returns the file where the signal strength/dominance values are saved in \&.csv format\&. 
.PP
\fBReturns\fP
.RS 4
an ofstream representing the file where the signal strength/dominance values for all antennas belonging to this mobile operator and all tiles of the grid are saved\&. 
.RE
.PP

.SS "const string MobileOperator::toString (bool detailed = \fCfalse\fP) const\fC [override]\fP, \fC [virtual]\fP"
Overrides the same method from the superclass\&. It is used to write the characteristics of the Mobile Operator to a file or to console\&. 
.PP
\fBParameters\fP
.RS 4
\fIdetailed\fP this parameter has no effect for the \fBMobileOperator\fP class\&. 
.RE
.PP
\fBReturns\fP
.RS 4
a string that describes the parameters of the MobieOperator\&. This string contains the id of the \fBMobileOperator\fP object and its name\&. 
.RE
.PP

.PP
Implements \fBAgent\fP\&.
.SS "void MobileOperator::writeSignalFileHeader ()"
Writes the header of the file where the signal strength/dominance is saved\&. 
.SH "Member Data Documentation"
.PP 
.SS "ofstream MobileOperator::m_antennaCellsFileName\fC [private]\fP"

.SS "const string MobileOperator::m_name\fC [private]\fP"

.SS "const string& MobileOperator::m_outputDir\fC [private]\fP"

.SS "const double MobileOperator::m_probMobilePhone\fC [private]\fP"

.SS "ofstream MobileOperator::m_signalMeasureFileName\fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for Simulator from the source code\&.


src/agent/MobileOperator.cpp
/*
 * Copyright (C) 2019  Bogdan Oancea

 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version and under the EUPL free software license version 1.0 or later.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/> and
 * <https://ec.europa.eu/info/european-union-public-licence_en>
 *
 * A data simulator for mobile phone network events
 *
 * MobileOperator.cpp
 *
 *  Created on: Jul 2, 2019
 *      Author: Bogdan.Oancea
 *      Email : bogdan.oancea@gmail.com
 */

#include <agent/MobileOperator.h>
#include <Constants.h>
#include <map/Map.h>
#include <iomanip>
#include <iostream>
#include <sstream>

using namespace std;

MobileOperator::MobileOperator(const Map* m, const unsigned long id, const Clock* clock, const char* name, const double probMobilePhone, const string& outputDir) :
		Agent(m, id, clock), m_name { name }, m_probMobilePhone { probMobilePhone }, m_outputDir{outputDir} {

	ostringstream cells;
	char sep = Constants::sep;
	cells << outputDir << "/"<< "AntennaCells_" << name << ".csv";
	try {
		m_antennaCellsFileName.open(cells.str(), ios::out);
	} catch (std::ofstream::failure& e) {
		cerr << "Error opening antenna cells output file!" << endl;
	}
	m_antennaCellsFileName << "AntennaId" << sep << "Cell Coordinates" << endl;

	ostringstream quality;
	quality << outputDir << "/"<< "SignalMeasure_" << name << ".csv";
	try {
		m_signalMeasureFileName.open(quality.str(), ios::out);
	} catch (std::ofstream::failure& e) {
		cerr << "Error opening signal strength/quality output file!" << endl;
	}
}

MobileOperator::~MobileOperator() {
	if (m_antennaCellsFileName.is_open()) {
		try {
			m_antennaCellsFileName.close();
		} catch (std::ofstream::failure& e) {
			cerr << "Error closing antenna cells output files!" << endl;
		}
	}
	if (m_signalMeasureFileName.is_open()) {
		try {
			m_signalMeasureFileName.close();
		} catch (std::ofstream::failure& e) {
			cerr << "Error closing signal quality output files!" << endl;
		}
	}
}

const string MobileOperator::toString(bool detailed) const {
	ostringstream result;
	result << left << setw(15) << getId() << setw(15) << getMNOName();
	return (result.str());
}

const string MobileOperator::getMNOName() const {
	return (m_name);
}

const double MobileOperator::getProbMobilePhone() const {
	return (m_probMobilePhone);
}

ofstream& MobileOperator::getAntennaCellsFile() {
	return (m_antennaCellsFileName);
}

ofstream& MobileOperator::getSignalFile() {
	return (m_signalMeasureFileName);
}


const string MobileOperator::getHeader() {
	ostringstream result;
	result << left << setw(15) << "MNO ID" << setw(15) << " Name " << endl;
	return (result.str());
}

void MobileOperator::writeSignalFileHeader() {
	char sep = Constants::sep;
	unsigned long noTiles = getMap()->getNoTiles();
	m_signalMeasureFileName << "Antenna ID" << sep;
	for (unsigned long i = 0; i < noTiles - 1; i++) {
		m_signalMeasureFileName << "Tile" << i << sep;
	}
	m_signalMeasureFileName << "Tile" << noTiles - 1 << endl;
}

const string& MobileOperator::getOutputDir() const {
	return m_outputDir;
}

